### PR TITLE
Light eater can no longer eat space tiles

### LINKED ISSUE #78318: Nightmare Light Eaters can darken space tiles
## Testmerges:

Server-Side Code Modifications: https://tgstation13.org/parsed-logs/sybil/server_side_modifications.dm

Round ID: _Bleep_
BYOND Version: 515.1610
Server revision compiled on: 
Master commit: [b5e346b2c42090f88fa02af311fba1a20a7d5b8d](https://www.github.com/tgstation/tgstation/commit/b5e346b2c42090f88fa02af311fba1a20a7d5b8d)
The following pull requests are currently test merged:
[#78217: 'Pipe gas visuals' by ninjanomnom at commit 63053a67b0](https://www.github.com/tgstation/tgstation/pull/78217)
[#77307: 'Revert "Bumps rust-g to `3.0.0` "' by MrStonedOne at commit 8dad0feceb](https://www.github.com/tgstation/tgstation/pull/77307)

Local commit: f4844d537b67e873a573b2d0d45fe24137a48b62
TGS version: /tg/station 13 Server v3.2.6.2
DMAPI version: 3.2.1.3

Current Informational Settings:
Protect Authority Roles From Traitor: 1
Protect Assistant Role From Traitor: 0
Enforce Human Authority: 1
Allow Latejoin Antagonists: 1

## Reproduction:
Be a nightmare, swing your light eater at a space tile:

nightmare_space_tile

### PR DESCRIPTION
**Pull request by Momo8289**
## About The Pull Request
This PR makes it so that the nightmare's light eater can't eat space tiles because well, it's space. The way I did it was just an explicit check to exclude space tiles specifically, as I can't really think of anything else it would apply to. Let me know if I should make it more general. Doesn't use `COMPONENT_BLOCK_LIGHT_EATER` because as mentioned by @MrMelbert, there's a lot of space tiles and that would be expensive
## Why It's Good For The Game
It's called light eater, not star eater. 
Fixes #78318
## Changelog
:cl:
fix: The Nightmare's Light Eater can no longer suck the light out of space tiles.
/:cl:

### CODE REVIEW

File: code/datums/elements/light_eater.dm [hunk 1]
```diff
@@ -51,7 +51,9 @@
 
 	. = 0
 	for(var/morsel in buffet)
-		. += devour(morsel, eater)
+		// You can't eat space, silly
+		if(!istype(morsel, /turf/open/space))
```
Comment (san7890) [inline code/datums/elements/light_eater.dm:55]: don't we have an `isspaceturf` or something check

Comment (Momo8289) [reply code/datums/elements/light_eater.dm:55]: Maybe but doesn't really matter, the problem isn't exclusive to space tiles and it's a whole bigger thing, working on it

**Review by MrMelbert — CHANGES_REQUESTED** [review]
I think it'd be better to put this check in `/datum/element/light_eater/proc/devour`, maybe as a typecache `undevourable_items = typecacheof(list(/turf/open/space))`. So we can expand it for other problem objects. 

The proper proper solution would be to have space turfs register `COMSIG_LIGHT_EATER_ACT` and return `COMPONENT_BLOCK_LIGHT_EATER` but given space is kind of expensive I'm not sure if it's worth it

**Comment by Momo8289** [comment]
> I think it'd be better to put this check in `/datum/element/light_eater/proc/devour`, maybe as a typecache `undevourable_items = typecacheof(list(/turf/open/space))`. So we can expand it for other problem objects.
> 
> The proper proper solution would be to have space turfs register `COMSIG_LIGHT_EATER_ACT` and return `COMPONENT_BLOCK_LIGHT_EATER` but given space is kind of expensive I'm not sure if it's worth it

Turns out this goes beyond just space tiles, looks like the whole system in place to let things be immune/have custom behaviour for light eater just doesn't work

**Comment by Momo8289** [comment]
> I think it'd be better to put this check in `/datum/element/light_eater/proc/devour`, maybe as a typecache `undevourable_items = typecacheof(list(/turf/open/space))`. So we can expand it for other problem objects.
> 
> The proper proper solution would be to have space turfs register `COMSIG_LIGHT_EATER_ACT` and return `COMPONENT_BLOCK_LIGHT_EATER` but given space is kind of expensive I'm not sure if it's worth it

Alright, gonna go with this and make another PR to address some of the other issues I've found with light eater

**Comment by the-orange-cow** [comment]
Nightmare flexes on Narsie, she wishes she could snuff out lights this easily.

File: code/datums/elements/light_eater.dm [hunk 2]
```diff
@@ -83,6 +83,9 @@
  * - [eater][/datum]: The light eater eating the morsel. This is the datum that the element is attached to that started this chain.
  */
 /datum/element/light_eater/proc/devour(atom/morsel, datum/eater)
+	var/undevourable = typecacheof(list(/turf/open/space))
+	if(morsel in undevourable)
```
Comment (Ghommie) [inline code/datums/elements/light_eater.dm:87]:
```suggestion
	var/static/list/undevourable = typecacheof(list(/turf/open/space))
	if(is_type_in_typecache(morsel, undevourable))
```

**Pull request merged**